code/modules/power/smes.dm

```diff
@@ -116,20 +118,18 @@
 /obj/machinery/power/smes/proc/chargedisplay()
 	return round(5.5*charge/(capacity ? capacity : 5e6))
 
-/obj/machinery/power/smes/proc/input_power(var/percentage)
+/obj/machinery/power/smes/proc/input_power(var/percentage, var/obj/machinery/power/terminal/term)

```

Comment:
That'll do it!  Only drawing from the specific terminal (combined with looping over terminals instead of SMES) will ensure the `percentage` parameter is applied only to the powernet that generated it.